### ISSUE TITLE
How to handle custom aspect ratios or multi-dimensional inputs for training

### ISSUE THREAD

**Issue by loucif01** [ultralytics/ultralytics#6880]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

I want to train YOLOv8 with the original size of my images, such as 1024x720, to maintain their quality. Using square sizes like 800x800 or 640x640 alters the aspect ratio and may degrade the image quality. How can I solve this?

**Comment by glenn-jocher** [comment]
@loucif01 hello! Training YOLOv8 with non-square image resolutions (like your 1024x720) is perfectly feasible. YOLOv8 is designed to handle various aspect ratios and will automatically manage the resizing for you while maintaining the aspect ratio of your original images during training. This means your images' quality and aspect ratio will be respected throughout the process. 😊

Keep in mind that your dataset's YAML file should specify the correct `width` and `height` to match your desired input resolution. The model will be trained on this specified resolution, ensuring the aspect ratios of your images are preserved.

For more details on how to set up your custom resolution in your data YAML, please refer to our documentation at the [Train](https://docs.ultralytics.com/modes/train) mode. Happy training! 🚀

**Comment by itanab** [comment]
Hi Glenn, I had a followup question. 

So to clarify,YOLO will automatically resize (keeping the aspect ratio) our images to 640x640 for training is that correct? 
So either we can either pre-process them offline to 640x640 or input them raw and let YOLO handle it?

Is 640x640 the only resolution that the model can accept?

**Comment by glenn-jocher** [comment]
@itanab hello! YOLOv8 will indeed automatically resize images to the resolution specified in your dataset's YAML file, while maintaining the original aspect ratio. You don't necessarily have to pre-process your images to 640x640; the model can handle raw images directly.

640x640 is not the only resolution accepted by YOLOv8. You can train the model with different resolutions by adjusting the `width` and `height` in your YAML configuration. The model is versatile and can be trained on various resolutions that best fit your needs. Happy training! 🌟

**Comment by loucif01** [comment]
@glenn-jocher Thanks for your help. Do we need to specify the custom size in the imgsz argument of the CLI, or is setting the **width** and **height** in the YAML file sufficient? I set the **width** and **height** in the YAML file, but the model didn't use these dimensions and defaulted to the size of 640!

**Comment by glenn-jocher** [comment]
@loucif01 hello there! If you've set the **width** and **height** in the YAML file but the model defaults to 640, make sure you're not overriding these with the `--imgsz` argument in the CLI. The settings in the YAML should be sufficient, but they need to be matched or omitted in the CLI to take effect. Check your training command and adjust as necessary. If the issue persists, ensure there are no conflicting settings in your setup. 👍

**Comment by vishalkumardas** [comment]
@glenn-jocher  I have a requirement to handle multiple Aspect Ratio textract OCRs of various invoices which have the coordinates normalized. We don't know the original layout. All we have is the OCR coordinates. We managed to build a pipeline that predicts certain fields from the OCR. We are using a Fully Convolution Network which is UNET. Now to improve accuracies, we are thinking about Aspect Ratios of the layout. I want to know how we can train different Aspect Ratio inputs without doing any resizing or padding on a UNET model.

**Comment by glenn-jocher** [comment]
@vishalkumardas, for training UNET models on OCR data with varying aspect ratios without resizing or padding, the UNET architecture should intrinsically handle multi-dimensional inputs due to its fully convolutional nature. You'll need to ensure your UNET implementation can accept variable-sized inputs and adjust the data loader accordingly. During training, batch sizes may need to be set to one to accommodate different image dimensions. Note that our YOLOv8 repo focuses on object detection tasks and while YOLO models also handle variable aspect ratios, they differ in architecture from UNET which is tailored for segmentation tasks. 😊

**Comment by vishalkumardas** [comment]
@glenn-jocher  Thanks for your reply. So to handle the multidimensional inputs, should I do multi-pathways for a single model like creating branches and allow each input to be fed of different dimensions?  Note: I don't want to resize or padding at any stage of this process.

**Comment by glenn-jocher** [comment]
@vishalkumardas, multi-pathway branches for different dimensions are not necessary. Since UNET is fully convolutional, it can inherently manage variable input sizes. Just ensure that your input pipeline and model are configured to handle varying dimensions without resizing or padding. You may also need to use a batch size of one if you encounter issues batching images of different sizes. Keep the processing consistent and avoid resizing/padding within the model for best results with your aspect ratios. 😊✨

**Comment by vishalkumardas** [comment]
@glenn-jocher Thanks again. I am using batch size = 1 to handle each input. But, I didn't understand, for example, I have 3 different resolutions. How will my model understand 3 different inputs, without resizing or padding.  I have tried that, It gave me error during the model definition. That's why I created 3 different pathways to incorporate my resolution-specific model. I know I might have missed something. Can you please help me out here, what should I use in my input pipeline to avoid multipath?

**Comment by vishalkumardas** [comment]
@glenn-jocher  Yes its working with batch size = 1, after doing some changes. The only problem I am facing right now is to save the best model with target sizes. But the problem is with saving it with different target sizes which is equivalent to the resolutions. I am posting some parts of my training code here. 

with mlflow.start_run(experiment_id=mf_logger.experiment_id, run_id=mf_logger.run_id) as run:
        save_source_code_artifacts()
        trainer.fit(seg_trainer, dm_internal)

        # Give combined statistics for all datasets
        trainer.test(ckpt_path='best', datamodule=dm_internal)

        # Save best model in MLFow using ONNX flavor
        onnx_model_path = Path('onnx_model.onnx')
        model.load_from_checkpoint(trainer.checkpoint_callback.best_model_path)  # type: ignore

        dummy_input = torch.zeros(1, n_input_channels, *target_size)
        dynamic_axes = {'input': {0: 'batch'}, 'output': {0: 'batch'}}
        torch.onnx.export(model, dummy_input, onnx_model_path, input_names=['input'], output_names=['output'],
                          dynamic_axes=dynamic_axes)

        onnx_model = onnx.load(onnx_model_path.as_posix())
        onnx.checker.check_model(onnx_model)

        for key, value in processor.serialize().items():
            md = onnx_model.metadata_props.add()
            md.key = key
            md.value = value

        mlflow.onnx.log_model(onnx_model, "model")

        onnx_model_path.unlink()
Here you can see, that the target size is required to save a model. I am confused if we are going to save 3 separate Onnx models with resolutions. Please share your thoughts on this. :)

**Comment by glenn-jocher** [comment]
@vishalkumardas, great to hear that your model is now training with batch size = 1. Regarding the ONNX export, you can actually save a single ONNX model that supports multiple resolutions. The key is to specify dynamic axes during export, which allows for variable input sizes. In your code snippet, you've already set 'dynamic_axes' appropriately. This means you don't need to save separate models for each resolution. Instead, export one ONNX model with dynamic axes, and it should handle varying input shapes during inference. Just ensure that your input preprocessing during inference matches the training conditions. 😊👍

**Comment by vishalkumardas** [comment]
@glenn-jocher  Thanks for the thought. Training and inference parts are successfully done with batch size =1 and I was able to save the best model parameters. However, when converting the pytorch lightning model to ONNX it gives me error because of dummy input. As UNET is a fully convolutional network and it doesn't takes an input, while passing 3 different input as a dummy to the export ONNX, it gives me the error as the model takes 1 argument, but passed 3. How to give only one dummy input.

        dummy_input1 = torch.zeros(1, n_input_channels, 656, 512)
        dummy_input2 = torch.zeros(1, n_input_channels, 864, 256)
        dummy_input3 = torch.zeros(1, n_input_channels, 512, 656)

        dummy_input = [dummy_input1, dummy_input2, dummy_input3]
        input_names = [ "dummy_input1", "dummy_input2", "dummy_input3"]
        dynamic_axes = {'input': {0: 'batch'}, 'output': {0: 'batch'}}
        torch.onnx.export(model, dummy_input, onnx_model_path, input_names=input_names, output_names=['output']),
                      dynamic_axes=dynamic_axes)

        onnx_model = onnx.load(onnx_model_path.as_posix())
        onnx.checker.check_model(onnx_model)

**Comment by glenn-jocher** [comment]
@vishalkumardas, when exporting to ONNX with dynamic axes for variable input dimensions, you don't need to provide multiple dummy inputs. You should create a single dummy input with unspecified dimensions for the height and width, using None to indicate variability. This will define the model to accept any input shape that agrees with the dynamic axes:

```python
dummy_input = torch.zeros(1, n_input_channels, None, None)
dynamic_axes = {'input': {0: 'batch', 2: 'height', 3: 'width'}, 'output': {0: 'batch'}}
torch.onnx.export(model, dummy_input, onnx_model_path, input_names=['input'], output_names=['output'], dynamic_axes=dynamic_axes)
```

This should create an ONNX model capable of handling inputs of different resolutions as per the dynamic axes specifications.

**Comment by vishalkumardas** [comment]
@glenn-jocher Thanks again. But torch doesn't support None as dimensions.

**Comment by glenn-jocher** [comment]
@vishalkumardas, apologies for the confusion. Instead of using `None`, you can actually use a variable shape in the dummy input and specify `dynamic_axes` to allow different resolutions. For ONNX export, choose an arbitrary resolution that suits your needs for the dummy_input, like so:

```python
dummy_input = torch.zeros(1, n_input_channels, height, width)  # Use a common resolution
dynamic_axes = {'input': {0: 'batch', 2: 'height', 3: 'width'}, 'output': {0: 'batch', 1: 'height', 2: 'width'}}
torch.onnx.export(model, dummy_input, onnx_model_path, input_names=['input'], output_names=['output'], dynamic_axes=dynamic_axes)
```

Remember that `height` and `width` are just placeholders for the arbitrary resolution you choose for the dummy input. The key is in `dynamic_axes`, which specifies that the model can accept dynamic sizes for batch, height, and width during inference. 😊🔧

**Comment by vishalkumardas** [comment]
@glenn-jocher Thanks. I will implement this, any tips to increase the training speed? It would be very slow when using batch size=1 for 4 million training samples.  The prerequisites for training are Preprocessing steps where we are doing OCRFlatter, CoordinatesNormalizer to relative position, Word2Char to convert texts at each coordinate to characters, ToGrid(involves creating chargrid of a certain size, and this is the process where we are leveraging our Aspect Ratio resolutions to create a grid of the same resolution size), OneHotEncoder and lastly converting it ToTensor. For postprocessing we are converting it back to Numpy after segmentation, then ArgMaz calculation and MaskToText
MaskToText - This method combines the OCR data and the segmentation mask, to extract the text that is on the mask. The first step is to build a word grid based on the OCR data. Then, it will calculate the intersection of both grids (word grid vs prediction grid). 
These are all the steps during inference and finally, we are calculating inference. We have managed to create our model to multitask and predict certain fields based on requirements.  We are using PyTorch Lightning data loader with a custom coalate function to handle tensor stacks during training. And defined a trainer to do the segmentation from the same. 
 Please suggest some ideas to increase the training speed.

**Comment by glenn-jocher** [comment]
@vishalkumardas, to enhance training speed with such a large dataset, consider the following strategies:

1. **Distributed Training**: Utilize multiple GPUs via Data Parallelism or Distributed Data Parallelism if available to distribute the workload.
2. **Mixed Precision Training**: Take advantage of mixed precision training to speed up computations and reduce memory usage on compatible hardware.
3. **Advanced Data Loading**: Enhance your dataloader with techniques such as pre-fetching, multi-threaded loading, and reducing image resolution if high resolution is not required for model performance.
4. **Optimize Preprocessing**: Streamline preprocessing steps. For example, merge certain processing functions to minimize the number of operations.
5. **Profile and Optimize**: Use profiling tools like PyTorch Profiler to identify bottlenecks in both training and data loading pipeline, and target those areas for optimization.
6. **Learning Rate Scheduler**: Employ a learning rate scheduler to potentially reduce the number of epochs needed for convergence.

Applying these techniques can help in addressing the slow training issue given your comprehensive batch size and large dataset setting. 😊🚀

**Comment by vishalkumardas** [comment]
@glenn-jocher Hey, thanks for the reply & sorry for the late updates. I trained my model with a single batch size and it was way too slow, the problem I faced in terms of training was that there were many noises at each step which resulted in poor convergence. Now we are sticking to our original approach. 
Initially, we had a requirement of handling single page OCR and now we will be handling multipage OCRs as well. The ratio of single-page to multipage OCRs is 80:20. In the case of a single page, there could be a scenario of multiple occurrences of certain fields on that page. Based on the highest confidence scores, we are predicting the fields. The same can be happening on multipage as well. What I am planning to deal with in such a situation is that I will split the pages before feeding it to model architecture(i.e. UNET) and handle individual pages and after inference, I will merge the confidences. 
However I am confused, how would I train a model in such cases, do you have any thoughts?

**Comment by glenn-jocher** [comment]
@vishalkumardas, for handling multipage OCRs, your approach to process individual pages separately and then aggregate the results based on confidence scores is sensible. For training, consider treating each page as an independent sample, even if they come from the same document. This way, your model learns to recognize fields on a per-page basis, which aligns with your inference strategy. Ensure your dataset includes a representative mix of single and multipage documents reflecting the 80:20 ratio. Additionally, augment your training data with scenarios where fields occur multiple times to teach the model to handle such cases. Keep the training consistent with the inference process for the best results. Good luck! 😊📄🔍

**Comment by cinzyy** [comment]
The model is trained with 640 imgsize but in input live stream sometimes the resolution is different resulting in different aspect ratio.. which causes bbox shift slightly downwards how to solve that..  Ive tried with maintaining target aspect ratio to 4:3 and add padding but yolo internally does that right? anyway after adding that part, the bbox is fitting fine but if yolo internally does that why do we have bbox shift problem??

**Comment by glenn-jocher** [comment]
YOLO does internally handle resizing and padding to maintain aspect ratio, but discrepancies can occur if the input preprocessing during inference isn't consistent with training. Ensure that your live stream frames undergo the same resizing and padding logic as during training to avoid bounding box shifts. If the issue persists, verify that the preprocessing pipeline aligns with YOLO's internal handling.

**Comment by cinzyy** [comment]
Thankyou for the reply!

**Comment by glenn-jocher** [comment]
You're welcome! If you have any further questions or need additional clarification, feel free to ask.

**Comment by cinzyy** [comment]
Hey, I'd appreciate your insights on the following..

1> I have trained yolov8 using roboflow with preprocessing techniques such as auto orient and image stretching.. should I add these pre processing before passing live stream frames to model ? 

2> I have trained my model with around 1855 images after augmentation in roboflow applying rotation 90 and exposure+10 -10 that come upto 5000 images is it enough? cause I have been getting error around +1 or -1 inch . I'm keeping ruler as reference in pictures to calculate scaling factor to predict dimension of parcel. Is it possible to get accurate results ?

3>I have ruler on top of parcel in images.
While annotating should I hover on the ruler if it is on parcel edge. 
Or 
Whether it should be annotated around the ruler.     (Does it affect performance?)

**Comment by glenn-jocher** [comment]
@cinzyy 1. Yes, if the preprocessing techniques like auto-orient and image stretching were used during training, you should apply the same preprocessing to live stream frames to ensure consistency and avoid discrepancies.

2. While 5000 augmented images might be sufficient in some cases, achieving high accuracy for precise measurements (+/-1 inch) can be challenging. Consider increasing the dataset size, diversifying the augmentations, or fine-tuning your model further with high-quality, annotated samples to improve dimension prediction accuracy.

3. Annotations should reflect the target object you aim to detect. If the ruler is on the parcel edge and part of your target, include it within the bounding box. However, if the ruler is not part of the intended object for detection, annotate only the parcel. Consistency in annotations is key to good performance.

**Comment by cinzyy** [comment]
Thank you very much!

**Comment by cinzyy** [comment]
Thankyou very much!

**Comment by glenn-jocher** [comment]
You're welcome! If you have any other questions or need further assistance, feel free to ask.

**Comment by cinzyy** [comment]
One more question: The model struggles with dark backgrounds. Should I train it with more dark-background images or adjust contrast instead?

**Comment by glenn-jocher** [comment]
To improve performance on dark backgrounds, training with more annotated images featuring dark backgrounds is recommended. This will help the model generalize better. Adjusting contrast during preprocessing might also help but is secondary to providing sufficient representative training data.

**Comment by cinzyy** [comment]
Thanks a lot

**Comment by cinzyy** [comment]
Hello!

I've been using the YOLOv8-seg model to predict the length and breadth of parcels. From what I understand, the accuracy largely depends on maintaining consistency between the training and inference environments and data. If the inference environment differs, we typically try to align it with the training conditions through preprocessing techniques. Is that correct?

I believe simply adding extra code won’t significantly improve accuracy. I’m uncertain about the best approach to achieve more accurate results. Currently, I’m still facing a +1 to -1 inch error for parcels ranging from 2 to 60 inches in size.

Any suggestions or guidance would be greatly appreciated! 😊

**Comment by glenn-jocher** [comment]
Yes, your understanding is correct; maintaining consistency between training and inference is crucial for accuracy. Since you're aiming for very high precision (+/- 1 inch), consider exploring techniques like knowledge distillation or quantization-aware training, as detailed in the Ultralytics documentation on [model deployment practices](https://docs.ultralytics.com/guides/model-deployment-practices/).

**Comment by cinzyy** [comment]
Thankyou !😊

**Comment by glenn-jocher** [comment]
You're welcome! If there's anything else you need assistance with, feel free to ask.

**Comment by cinzyy** [comment]
Hello, I trained my model using images of high resolution anyway i resized to 640 before passing to yolov8. I recently tried to inference the model with webcam images of size 640. but the detection is not as consistent as it was for those images. why? 
1> Should i include images captured by webcam ?
2> should i enhance image before passing to model or try pre processing.

**Comment by glenn-jocher** [comment]
@cinzyy The inconsistency between your high-resolution training images and webcam inference is likely due to domain shift - differences in image quality, lighting conditions, and sensor characteristics. To address this:

1. Yes, definitely include webcam-captured images in your training dataset. This is the most effective solution as it helps the model learn the specific characteristics of your deployment environment.

2. Apply basic preprocessing to webcam images such as normalization and color correction to make them more similar to your training distribution. You might also benefit from data augmentation during training with techniques that simulate webcam quality (noise addition, blur, etc.).

For best results, make your training data as representative as possible of your inference environment. The closer these match, the more consistent your detections will be. Check out the [data preprocessing guide](https://docs.ultralytics.com/guides/preprocessing_annotated_data/) for more techniques.

**Comment by cinzyy** [comment]
Thanks a ton for detailed explanation! I’ll follow your suggestions and start working on improving my dataset and preprocessing. Appreciate your help!😊

**Comment by glenn-jocher** [comment]
You're very welcome! I'm glad the explanation was helpful. Building a robust dataset with images from your actual deployment environment is indeed one of the most effective strategies. Looking forward to hearing about your progress with the improved dataset and preprocessing pipeline. Feel free to reach out if you encounter any new challenges along the way.
